[SPARK-5579][SQL][DataFrame] Support for project/filter using SQL expressions

rxin opened the pull request:
``` scala
df.selectExpr("abs(colA)", "colB")
df.filter("age > 21")
```

rxin committed:
[SPARK-5579][SQL][DataFrame] Support for project/filter using SQL expressions.

e.g.

df.selectExpr("abs(colA)", "colB")

df.filter("age > 21")

SparkQA commented:
[Test build #26693 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26693/consoleFull) for   PR 4348 at commit [`ac65f4b`](https://github.com/apache/spark/commit/ac65f4b86bfce0ba2e170c31f6a50c58255f960e).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26693 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26693/consoleFull) for   PR 4348 at commit [`ac65f4b`](https://github.com/apache/spark/commit/ac65f4b86bfce0ba2e170c31f6a50c58255f960e).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class Dsl(object):`
  * `class ExamplePointUDT(UserDefinedType):`
  * `class SQLTests(ReusedPySparkTestCase):`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26693/
Test FAILed.

rxin committed:
[SPARK-5578][SQL][DataFrame] Provide a convenient way for Scala users to use UDFs

A more convenient way to define user-defined functions.

Author: Reynold Xin <rxin@databricks.com>

Closes #4345 from rxin/defineUDF and squashes the following commits:

639c0f8 [Reynold Xin] udf tests.
0a0b339 [Reynold Xin] defineUDF -> udf.
b452b8d [Reynold Xin] Fix UDF registration.
d2e42c3 [Reynold Xin] SQLContext.udf.register() returns a UserDefinedFunction also.
4333605 [Reynold Xin] [SQL][DataFrame] defineUDF.

rxin committed:
Fix Python.

rxin commented on python/pyspark/sql.py:
```diff
@@ -2126,10 +2126,9 @@ def sort(self, *cols):
         if not cols:
             raise ValueError("should sort by at least one column")
-        jcols = ListConverter().convert([_to_java_column(c) for c in cols[1:]],
+        jcols = ListConverter().convert([_to_java_column(c) for c in cols],
```
@davies take a look at the Python changes.

SparkQA commented:
[Test build #26723 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26723/consoleFull) for   PR 4348 at commit [`2baeef2`](https://github.com/apache/spark/commit/2baeef2f4035bad7aa829cf52fc338245f52fafd).
 * This patch merges cleanly.

davies commented on sql/core/src/main/scala/org/apache/spark/sql/DataFrameImpl.scala:
```diff
@@ -179,10 +179,20 @@ private[sql] class DataFrameImpl protected[sql](
     select((col +: cols).map(Column(_)) :_*)
   }
 
+  override def selectExpr(exprs: String*): DataFrame = {
```
I think this one could be merged into select(), column is also a valid expression

rxin replied:
not if it has space ... it will just fail

davies replied:
It should work in these cases with this implementation.
```
select('a', '`the name`', 'a + 1', 'min(b) * 3')
```

rxin replied:
yea - but asking users to wrap a column name in backticks in strings is fairly annoying.

davies commented:
This select() and filter() in Python do not support yet

SparkQA commented:
[Test build #26723 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26723/consoleFull) for   PR 4348 at commit [`2baeef2`](https://github.com/apache/spark/commit/2baeef2f4035bad7aa829cf52fc338245f52fafd).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26723/
Test PASSed.

rxin commented:
We can discuss more offline. For now let's keep this separate, otherwise it can be fairly annoying to use column names that contain space or column names that contain any SQL keywords.
